server/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodeRole.java

```diff
@@ -242,4 +261,84 @@ public void validateRoles(final List<DiscoveryNodeRole> roles) {
 
     }
 
+    // the set of possible roles
+    private static final SortedSet<DiscoveryNodeRole> ROLES;
+
+    /**
+     * The possible node roles.
+     *
+     * @return an ordered, immutable set of possible roles
+     */
+    public static SortedSet<DiscoveryNodeRole> roles() {
+        return ROLES;
+    }
+
+    // a map from role names to their role representations
+    private static final Map<String, DiscoveryNodeRole> ROLE_MAP;
+
+    static {
+        final List<Field> roleFields = Arrays.stream(DiscoveryNodeRole.class.getFields())
+            .filter(f -> f.getType().equals(DiscoveryNodeRole.class))
+            .collect(Collectors.toUnmodifiableList());
+        // this will detect duplicate role names
+        final Map<String, DiscoveryNodeRole> roleMap = roleFields.stream()
+            .map(f -> {
+                try {
+                    return (DiscoveryNodeRole) f.get(null);
+                } catch (final IllegalAccessException e) {
+                    throw new AssertionError(e);
+                }
+            })
+            .collect(Collectors.toMap(DiscoveryNodeRole::roleName, Function.identity()));
+        assert roleMap.size() == roleFields.size() :
+            "roles by name [" + roleMap + "], role fields [" + roleFields + "]";
```

Comment:
I'm on the fence, but because this is using reflection and kind of scary, how about just making this an actual exception (even if it's an `AssertionError`) so that disabling assertions means duplicates still get caught?

Reply:
The lines before these would fail if these assertions are violated. This is because the JDK collectors don't let you have duplicate keys. So these assertions are really capturing/commenting/reflecting an assumption we are making about the previous lines. So if we refactor, this guard is here about behavior we're assuming. And if this assertion is ever going to fail, it's going to fail violently in tests and there's no avoiding that. It's only going to fail if we've made a really horrible mistake, like created two role instances with the same name, or the other assertion with the same abbreviation, and refactored how these maps are built. It's a safety check, to be 100% certain we didn't make a mistake, which is already unlikely to occur, since we're deliberate about adding new roles, and are aware of the names and abbreviations of our existing roles, and the production code already throws. That is, I truly see this an assertion, to make sure an assumption we're making actually holds, and not something that will ever be encounter in production. That's why I think it should be and remain an assertion.

Reply:
> This is because the JDK collectors don't let you have duplicate keys.

I didn't know this! Thanks for sharing!